[Serialization] Recover from a missing conforming type in a protocol conformance

xymus opened the pull request:
We saw this failure with a Clang module imported @_implementationOnly with synthesized conformances by the ClangImporter. It caused issues only in sil-opt as it reads all the witness tables.

rdar://problem/58924131

xymus committed:
[Serialization] Recover from a missing conforming type in a protocol conformance

We saw this failure with a Clang module imported @_implementationOnly
with synthesized conformances by the ClangImporter. It caused
issues only in sil-opt as it reads all the witness tables.

rdar://problem/58924131

xymus commented:
@swift-ci please smoke test

nkcsgexi commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -3108,8 +3118,12 @@ SILWitnessTable *SILDeserializer::readWitnessTable(DeclID WId,
 
   // Deserialize Conformance.
+  auto maybeConformance = MF->readConformanceChecked(SILCursor);
+  if (!maybeConformance)
```
could you add some comments why this error may happen? Otherwise LGTM.

xymus replied:
I've added documentation on the prototype for readWitnessTableChecked and on the only client readWitnessTables.

slavapestov reviewed (commented):
This needs a test case.

xymus committed:
[Serialization] Fix crash in reporting missing implementationOnly modules

The diagnostic function crashed as it did not take into account the
DebuggerSupport options as the site reporting the missing dependency did.
In this context, missing implementation-only imported dependencies are
ignored only if DebuggerSupport is set.

xymus committed:
[Test] Reproducer for witness table deserialization failure with a enum_flag

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

futurejones commented:
Can one of the admins verify this patch?

xymus commented:
@swift-ci please smoke test

xymus commented:
@swift-ci please test Windows platform
